proposals/multiplatform/binary-signature.md

```diff
@@ -0,0 +1,601 @@
+
+Currently, there is no way to work around this problem. For top-level functions on Kotiln/JVM it is customary to use 
+`@JvmName` for such a migration, but for other platforms or for open/abstract members no solution currently exists.
+
+The newly introduced `@BinarySignatureName` annotation can be used here with the following procedure for 
+all kinds of binary-compatible library migration scenarios:
+
+* The old function is marked with `@Deprecated(level = DeprecationLevel.HIDDEN)`, thus keeping its binary signature 
+  and supporting all previously compiled clients for this library on all platforms.
+* The new function is marked with `@BinarySignatureName("...")` giving a new name for the binary signature of this function,  
+  thus avoiding conflicting overloads, platform declaration clashes, and allowing to specify matching declaration for override.
+
+In the above example, the resulting code looks like this:
+
+```kotlin
+sealed class Base {
+    @Deprecated("Binary compatibility", level = DeprecationLevel.HIDDEN)
+    abstract fun doSomething()
+    
+    @BinarySignatureName("doSomethingString")
+    abstract fun doSomething(): String
+}
+
+class Derived : Base() { 
+     @Deprecated("Binary compatibility", level = DeprecationLevel.HIDDEN)
+     override fun doSomething()
+     
+     @BinarySignatureName("doSomethingString")
+     override fun doSomething(): String { /* ... */ }
+}
+```
+
+> The requirement to specify `@BinarySignatureName` on override is not convenient, so this solution is clearly not 
+> designed for everyday mass usage. Yet, it does solve J2K conversion problem while maintaining binary compatibility 
+> and avoids the problem of figuring out the platform name in the presence of multiple inheritance.
+
+### Use-case: Objective-C interoperability
+
+Objective-C uses _selectors_ to identify method calls. An Objective-C selector includes the method name and argument labels. 
+Objective-C can have methods in the same scope that differ only in the argument labels. 
+For example, consider the snippet from the following Objective-C protocol:
+
+```objective-c
+@protocol UITableViewDataSource
+- (nullable NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section;
+- (nullable NSString *)tableView:(UITableView *)tableView titleForFooterInSection:(NSInteger)section;
+@end
+```
+
+In Kotlin this protocol is represented with the following interface:
+
+```kotlin
+interface UITableViewDataSource { // ERROR: Conflicting overloads
+    fun tableView(tableView: UITableView, titleForHeaderInSection: Int): String?
+    fun tableView(tableView: UITableView, titleForFooterInSection: Int): String? 
+}
+```
+
+Normally, such overloads, that are different only in parameter names, are not allowed in Kotlin, so a special
+Objective-C specific rules are currently used for Objective-C interoperability. However, these Objective-C-specific 
```

Comment:
```suggestion
Normally, such overloads, that are different only in parameter names, are not allowed in Kotlin, so special
Objective-C specific rules are currently used for Objective-C interoperability. However, these Objective-C-specific 
```

An indefinite article is almost never used with plurals, it can be dropped